### PR TITLE
Support native injection for @Factory methods.

### LINKED ISSUE #1745: java.lang.IllegalArgumentException: wrong number of arguments
Regression from 6.11 to 6.14.2

```
The factory method class com.polo.primetest.factories.FactoryTest.createInstances() threw an exception
    at org.testng.internal.FactoryMethod.invoke(FactoryMethod.java:197)
    at org.testng.internal.TestNGClassFinder.processFactory(TestNGClassFinder.java:223)
    at org.testng.internal.TestNGClassFinder.processMethod(TestNGClassFinder.java:179)
    at org.testng.internal.TestNGClassFinder.processClass(TestNGClassFinder.java:171)
    at org.testng.internal.TestNGClassFinder.<init>(TestNGClassFinder.java:121)
    at org.testng.TestRunner.initMethods(TestRunner.java:370)
    at org.testng.TestRunner.init(TestRunner.java:271)
    at org.testng.TestRunner.init(TestRunner.java:241)
    at org.testng.TestRunner.<init>(TestRunner.java:167)
    at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:663)
    at org.testng.SuiteRunner.init(SuiteRunner.java:260)
    at org.testng.SuiteRunner.<init>(SuiteRunner.java:198)
    at org.testng.TestNG.createSuiteRunner(TestNG.java:1295)
    at org.testng.TestNG.createSuiteRunners(TestNG.java:1273)
    at org.testng.TestNG.runSuitesLocally(TestNG.java:1128)
    at org.testng.TestNG.runSuites(TestNG.java:1049)
    at org.testng.TestNG.run(TestNG.java:1017)
    at com.polo.primetest.testsupport.TestInitiator.createTestNG(TestInitiator.java:436)
    at com.polo.primetest.testsupport.TestInitiator.init(TestInitiator.java:302)
    at com.polo.primetest.execution.main.Main.main(Main.java:22)
Caused by: java.lang.IllegalArgumentException: wrong number of arguments
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.testng.internal.FactoryMethod.invoke(FactoryMethod.java:167)
    ... 19 common frames omitted
```

```java
public class FactoryTest
    {
        @Factory
        public Object[] createInstances( ITestContext ctx ) throws Exception
        {
            XmlTest currentXmlTest = ctx.getCurrentXmlTest();
            String parameterValue = currentXmlTest.getParameter( currentXmlTest.getName() );
        }
```

See https://stackoverflow.com/q/49570901/4234729

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1745

Fixes #1745 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Review by juherr — APPROVED** [review]

File: src/main/java/org/testng/internal/Parameters.java
Comment (cbeust) [inline src/main/java/org/testng/internal/Parameters.java]: Missing space before `xmlSuite`.

File: src/main/java/org/testng/internal/Parameters.java
Comment (cbeust) [inline src/main/java/org/testng/internal/Parameters.java]: No `final` on parameters (also not a big fan of the alignment since it causes more work when modifying the parameter list, but not a huge deal).

**Review by cbeust — COMMENTED** [review]
Feel free to merge after addressing the comments.

Thanks!

File: src/main/java/org/testng/internal/Parameters.java
Comment (krmahadevan) [reply src/main/java/org/testng/internal/Parameters.java]: Fixed.

File: src/main/java/org/testng/internal/Parameters.java
Comment (krmahadevan) [reply src/main/java/org/testng/internal/Parameters.java]: Thanks @cbeust That was copy paste error :) The previous method had the `final` keyword too. Cleaned up the final keyword usage in both the places.

**Comment by krmahadevan** [comment]
Waiting for the CI build to run to completion. Once its GREEN, will merge.

**Pull request merged**
